go/vt/vtgate/planbuilder/operators/horizon_phases.go

```diff
@@ -0,0 +1,196 @@
+)
+
+// Phase defines the different planning phases to go through to produce an optimized plan for the input query.
+type Phase struct {
+	Name string
+	// preOptimizeAction is the action to be taken before calling plan optimization operation.
+	preOptimizeAction func(ctx *plancontext.PlanningContext, op ops.Operator) (ops.Operator, error)
+	// preOptimizeAction is the action to be taken before calling plan optimization operation.
+	postOptimizeAction func(op ops.Operator) (ops.Operator, error)
+}
+
+// getPhases returns the phases the planner will go through.
+// It's used to control so rewriters collaborate correctly
+func getPhases() []Phase {
+	return []Phase{{
+		// Initial optimization
+		Name: "initial horizon planning optimization phase",
+	}, {
+		Name: "add columns if needed by filter",
+		preOptimizeAction: func(ctx *plancontext.PlanningContext, op ops.Operator) (ops.Operator, error) {
+			return addColumnsToFilter(ctx, op)
+		},
+	}, {
+		// Adding Ordering Op - Any aggregation that is performed in the VTGate needs the input to be ordered
+		Name: "add ORDER BY to aggregations above the route and add GROUP BY to aggregations on the RHS of join",
+		preOptimizeAction: func(ctx *plancontext.PlanningContext, op ops.Operator) (ops.Operator, error) {
+			ctx.Phases.PushAggregation = true
+			return addOrderBysForAggregations(ctx, op)
+		},
+		// Adding Group by - This is needed if the grouping is performed on a join with a join condition then
+		//                   aggregation happening at route needs a group by to ensure only matching rows returns
+		//                   the aggregations otherwise returns no result.
+		postOptimizeAction: addGroupByOnRHSOfJoin,
+	}}
+}
+
+func addColumnsToFilter(ctx *plancontext.PlanningContext, root ops.Operator) (ops.Operator, error) {
+	visitor := func(in ops.Operator, _ semantics.TableSet, isRoot bool) (ops.Operator, *rewrite.ApplyResult, error) {
+		filter, ok := in.(*Filter)
+		if !ok {
+			return in, rewrite.SameTree, nil
+		}
+
+		columns, err := filter.GetColumns()
+		if err != nil {
+			return nil, nil, vterrors.Wrap(err, "while planning filter")
+		}
+		proj, areOnTopOfProj := filter.Source.(selectExpressions)
+		if !areOnTopOfProj {
+			// not much we can do here
+			return in, rewrite.SameTree, nil
+		}
+		addedColumns := false
+		for _, expr := range filter.Predicates {
+			sqlparser.Rewrite(expr, func(cursor *sqlparser.Cursor) bool {
+				e, ok := cursor.Node().(sqlparser.Expr)
+				if !ok {
+					return true
+				}
+				offset := slices.IndexFunc(columns, func(expr *sqlparser.AliasedExpr) bool {
+					return ctx.SemTable.EqualsExprWithDeps(expr.Expr, e)
+				})
+
+				if offset >= 0 {
+					// this expression can be fetched from the input - we can stop here
+					return false
+				}
+
+				if fetchByOffset(e) {
+					// this expression has to be fetched from the input, but we didn't find it in the input. let's add it
+					_, addToGroupBy := e.(*sqlparser.ColName)
```

Comment:
add a comment explaining why we are doing this type test

Reply:
done